File: mcp/test/fixtures/bugfix-java/misses/EmptyCatch.java
public class EmptyCatch {
    void logs(String p) {
        try { Integer.parseInt(p); } catch (NumberFormatException e) { log(e); }
    }
    void rethrows(String p) {
        try { Integer.parseInt(p); } catch (NumberFormatException e) { throw new IllegalStateException(e); }
    }
    void log(Exception e) { }

    // The three below are DISCRIMINATING for the naming exclusion, and each
    // for one alternative of it: delete `ignored` from the
    // `metavariable-regex` on $V and only ignoredName fires, and so on.
    //
    // Semgrep matches the AST, and comments are not in it — a catch block
    // documented as a deliberate ignore is byte-identical to a silent
    // swallow, so the comment can never be the escape hatch. The variable
    // NAME can be, and it is already the convention two of the standard Java
    // linters use: Checkstyle's EmptyCatchBlock allows `ignore` and
    // `expected` by default, IntelliJ's inspection allows `ignored`. Naming
    // the exception one of those is the self-documenting way to say
    // "deliberate" without a suppression comment.
    //
    // Every other name still fires — the rule is not weakened, it is given a
    // vocabulary.
    void ignoredName(String p) {
        // IntelliJ's convention. Absence of a parsable value is expected here.
        try { Integer.parseInt(p); } catch (NumberFormatException ignored) { }
    }
    void ignoreName(String p) {
        // Checkstyle's first default.
        try { Integer.parseInt(p); } catch (NumberFormatException ignore) { }
    }
    void expectedName(String p) {
        // Checkstyle's second default: the throw IS the assertion.
        try { Integer.parseInt(p); } catch (NumberFormatException expected) { }
    }

    // The four below exist because the `finally` branch added to the positive
    // pattern could have been bolted on WITHOUT the naming escape hatch and
    // without the non-empty-body requirement, and nothing would have noticed:
    // before the fix every one of these was silent for the wrong reason — the
    // `finally` was doing the silencing, not the rule's own logic. They pin
    // that the two properties survive on the new branch as well as the old
    // one. Delete the `metavariable-regex` and the first two fire; write the
    // `finally` branch without requiring an EMPTY catch body and the last two
    // fire.
    void ignoredNameWithFinally(String p) {
        try { Integer.parseInt(p); } catch (NumberFormatException ignored) { } finally { cleanup(); }
    }
    void ignoredNameMultiCatch(String p) {
        try { Integer.parseInt(p); } catch (NumberFormatException | NullPointerException ignored) { }
    }
    void logsWithFinally(String p) {
        try { Integer.parseInt(p); } catch (NumberFormatException e) { log(e); } finally { cleanup(); }
    }
    void rethrowsWithFinally(String p) {
        try { Integer.parseInt(p); } catch (NumberFormatException e) { throw new IllegalStateException(e); } finally { cleanup(); }
    }

    void cleanup() { }
}

File: mcp/test/fixtures/bugfix-cs/misses/EmptyCatch.cs
// NEAR-MISSES for bugfix-cs-error-handling-empty-catch.
//
// Written BEFORE hits/EmptyCatch.cs. If any method here fires, the rule is
// wrong.
//
// The three naming exemptions are the rule's only escape hatch, and it is not
// a guard: it is a DECLARATION OF INTENT the rule itself reads. Semgrep
// matches the AST and comments are not in it, so a catch documented as a
// deliberate ignore is byte-identical to a silent swallow — the comment could
// never be the hatch. The name can be, and it is already the convention of
// the standard linters: Checkstyle's `ignore`/`expected`, IntelliJ's
// `ignored`. Roslyn has no equivalent default, so the vocabulary is inherited
// from the Java pack rather than from a C# tool.
using System;

namespace Guardian.Fixtures.Misses;

public sealed class EmptyCatch
{
    // The failure is recorded.
    public void Logs(string p)
    {
        try { Parse(p); } catch (FormatException ex) { Log(ex); }
    }

    // The failure is rethrown wrapped.
    public void Rethrows(string p)
    {
        try { Parse(p); } catch (FormatException ex) { throw new InvalidOperationException("bad input", ex); }
    }

    // The three naming conventions, one method each, and each is
    // DISCRIMINATING for one alternative of the `metavariable-regex`: drop
    // `ignored` from it and only IgnoredName fires, and so on.
    public void IgnoredName(string p)
    {
        // IntelliJ's convention. A non-numeric string is expected here.
        try { Parse(p); } catch (FormatException ignored) { }
    }

    public void IgnoreName(string p)
    {
        // Checkstyle's first default.
        try { Parse(p); } catch (FormatException ignore) { }
    }

    public void ExpectedName(string p)
    {
        // Checkstyle's second default: the throw IS the assertion.
        try { Parse(p); } catch (FormatException expected) { }
    }

    // An exception FILTER whose body logs. The filter is the shape most
    // likely to confuse a `catch ($E $V) { }` pattern, because the `when`
    // clause sits between the parameter and the block.
    public void FilteredAndLogged(string p)
    {
        try { Parse(p); } catch (FormatException ex) when (p.Length > 0) { Log(ex); }
    }

    // try/finally with NO catch clause at all. Nothing is swallowed — the
    // exception propagates.
    public void FinallyOnly(string p)
    {
        try { Parse(p); } finally { Cleanup(); }
    }

    // The same, with an EMPTY finally block. This is the near-miss for the
    // bare `catch { }` branch specifically: the token before the empty block
    // is the only thing distinguishing the two, so a branch written as
    // `try { ... } { }` — or one that matched any trailing empty block —
    // would fire here.
    public void EmptyFinallyBlock(string p)
    {
        try { Parse(p); } finally { }
    }

    private static void Parse(string p) => int.Parse(p);

    private static void Log(Exception ex) => Console.Error.WriteLine(ex.Message);

    private static void Cleanup() { }
}
